ext/opcache/jit/zend_jit_trace.c

```diff
@@ -0,0 +1,3972 @@
+			zend_jit_check_timeout(&dasm_state, opline);
+		}
+	}
+
+	ssa_op = ssa->ops;
+	for (;;p++) {
+		if (p->op == ZEND_JIT_TRACE_VM) {
+			uint8_t op1_type = p->op1_type;
+			uint8_t op2_type = p->op2_type;
+			uint8_t op3_type = p->op3_type;
+
+			opline = p->opline;
+			if (op1_type & IS_TRACE_REFERENCE) {
+				op1_type = IS_UNKNOWN;
+			}
+			if (op2_type & IS_TRACE_REFERENCE) {
+				op2_type = IS_UNKNOWN;
+			}
+			if (op3_type & IS_TRACE_REFERENCE) {
+				op3_type = IS_UNKNOWN;
+			}
+
+			if ((p+1)->op == ZEND_JIT_TRACE_OP1_TYPE) {
+				// TODO: support for recorded classes ???
+				p++;
+			}
+			if ((p+1)->op == ZEND_JIT_TRACE_OP2_TYPE) {
+				// TODO: support for recorded classes ???
+				p++;
+			}
+
+#if 0
+			// TODO: call level calculation doesn't work for traces ???
+			switch (opline->opcode) {
+				case ZEND_INIT_FCALL:
+				case ZEND_INIT_FCALL_BY_NAME:
+				case ZEND_INIT_NS_FCALL_BY_NAME:
+				case ZEND_INIT_METHOD_CALL:
+				case ZEND_INIT_DYNAMIC_CALL:
+				case ZEND_INIT_STATIC_METHOD_CALL:
+				case ZEND_INIT_USER_CALL:
+				case ZEND_NEW:
+					call_level++;
+			}
+#endif
+
+			if (zend_jit_level >= ZEND_JIT_LEVEL_INLINE) {
+				switch (opline->opcode) {
+					case ZEND_PRE_INC:
+					case ZEND_PRE_DEC:
+					case ZEND_POST_INC:
+					case ZEND_POST_DEC:
+						if (opline->op1_type != IS_CV) {
+							break;
+						}
+						op1_info = OP1_INFO_EX();
+						CHECK_OP1_TRACE_TYPE();
+						if (!(op1_info & MAY_BE_LONG)) {
+							break;
+						}
+						if (opline->result_type != IS_UNUSED) {
+							res_use_info = RES_USE_INFO_EX();
+							USE_RES_TRACE_TYPE();
+							res_info = RES_INFO_EX();
+							res_addr = RES_REG_ADDR();
+						} else {
+							res_use_info = -1;
+							res_info = -1;
+							res_addr = 0;
+						}
+						op1_def_info = OP1_DEF_INFO_EX();
+						if (!zend_jit_inc_dec(&dasm_state, opline, op_array,
+								op1_info, OP1_REG_ADDR(),
+								op1_def_info, OP1_DEF_REG_ADDR(),
+								res_use_info, res_info,
+								res_addr,
+								(op1_def_info & MAY_BE_LONG) && (op1_def_info & (MAY_BE_DOUBLE|MAY_BE_GUARD)) && zend_may_overflow_ex(opline, ssa_op, op_array, ssa))) {
+							goto jit_failure;
+						}
+						if ((op1_def_info & (MAY_BE_ANY|MAY_BE_GUARD)) == (MAY_BE_LONG|MAY_BE_GUARD)) {
+							ssa->var_info[ssa_op->op1_def].type &= ~MAY_BE_GUARD;
+							if (opline->result_type != IS_UNUSED) {
+								ssa->var_info[ssa_op->result_def].type &= ~MAY_BE_GUARD;
+							}
+						}
+						goto done;
+					case ZEND_BW_OR:
+					case ZEND_BW_AND:
+					case ZEND_BW_XOR:
+					case ZEND_SL:
+					case ZEND_SR:
+					case ZEND_MOD:
+						op1_info = OP1_INFO_EX();
+						CHECK_OP1_TRACE_TYPE();
+						op2_info = OP2_INFO_EX();
+						CHECK_OP2_TRACE_TYPE();
+						if ((op1_info & MAY_BE_UNDEF) || (op2_info & MAY_BE_UNDEF)) {
+							break;
+						}
+						if (!(op1_info & MAY_BE_LONG)
+						 || !(op2_info & MAY_BE_LONG)) {
+							break;
+						}
+						if (opline->result_type == IS_TMP_VAR
```

Comment:
This is a kind of peephole optimization that was implemented in method JIT early. In tracing JIT I just reuse it. The optimization merges operation with the following SEND to avoid double copying.